cascalog-core/src/clj/cascalog/logic/parse.clj

```diff
@@ -642,6 +642,32 @@ This won't work in distributed mode because of the ->Record functions."
         (assoc :available-fields fields)
         (assoc :ground? (v/fully-ground? fields)))))
 
+(defn prune-operations
+  "Remove non-generator & non-filter operations whose outvar(s) is not used, i.e., when the outvar(s)
+  do *not* intersect with the query's out-fields, generators outvars (for joins), invars of other operations,
+  and :sort option. Additionally, do not prune operations if a no-input operator exists"
+  [fields grouped options]
+  (let [check-input-operations (concat (grouped Operation)
+                                       (grouped FilterOperation)
+                                       (grouped Aggregator))]
+  (if (some #(empty? (:input %)) check-input-operations)
+    (concat (grouped Operation)
+            (grouped FilterOperation))
+    (let [generators (concat (grouped Generator)
+                             (grouped GeneratorSet))
+          necessary-fields (->> check-input-operations
+                                (mapcat :input)
+                                (concat fields)
+                                (concat (mapcat :fields generators))
+                                (concat (:sort options)))
+          pruned-operations (filter #(seq
+                                       (clojure.set/intersection
+                                            (set (:output %))
+                                            (set necessary-fields)))
+                                    (grouped Operation))]
+      (concat pruned-operations
+              (grouped FilterOperation))))))
```

Comment:
return pruned-operations and filter operations. so basically, filter operations don't even get touched.